File: Question 0040 - Combination Sum II/src/Solution.java
import java.util.*;

class Solution {
    private void solve(int[] candidates, int target, List<Integer> combi, int index, List<List<Integer>> results)
    {
        if (target == 0)
        {
            results.add(new ArrayList<>(combi));
            return;
        }

        for (int i = index; i < candidates.length; i++)
        {
            // Skip duplicates
            if (i > index && candidates[i] == candidates[i - 1]) continue;

            // Check if current candidate can be included
            if (candidates[i] <= target)
            {
                combi.add(candidates[i]);
                solve(candidates, target - candidates[i], combi, i + 1, results); // Pass i+1 to avoid reusing same element
                combi.removeLast(); // Backtrack
            }
        }
    }

    public List<List<Integer>> combinationSum2(int[] candidates, int target)
    {
        Arrays.sort(candidates);
        List<List<Integer>> results = new ArrayList<>();
        solve(candidates, target, new ArrayList<>(), 0, results);
        return results;
    }
}

File: Question 0040 - Combination Sum II/src/Solution.cpp
#include <bits/stdc++.h>
using namespace std;
#define pb push_back
#define vvi vector<vector<int>>
#define vi vector<int>

void solve(vi& candidates, vvi& ans, vi& combi, int target, int idx)
{
    if(target == 0)
    {
        ans.pb(combi);
        return;
    }

    for(int i = idx; i < candidates.size(); i++)
    {
        if(i > idx && candidates[i] == candidates[i-1]) continue;

        if(candidates[i] <= target)
        {
            combi.pb(candidates[i]);
            solve(candidates, ans, combi, target-candidates[idx], idx+1);
            combi.pop_back();
        }
    }
    
}

class Solution 
{
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) 
    {
        vvi ans;
        vi combi;
        sort(candidates.begin(), candidates.end());
        solve(candidates, ans, combi, target, 0);
        return ans;
    }
};